### ISSUE TITLE
Improve error message

### ISSUE THREAD

**Issue by jo-so** [emacs-lsp/lsp-mode#2395]
**Describe the bug**

Using Rustic and lsp-mode for rust files gives me *sometimes* the message `LSP :: Error from the Language Server: Rust Analyzer is still loading... (Unknown error)` This is not very helpful.

**To Reproduce**

I don't have a clear way to reproduce it, I only open some rust files and sometimes the message comes, often not.

**Expected behavior**

It would be more helpful to get a message `LSP :: Error from the Language Server: Rust Analyzer returned an unknown response to request 5. (Enable lsp-log-io)`.

More details should go to \*lsp-log\*. The message could be `LSP :: Error from the Language Server: Rust Analyzer returned an unknown response; see *lsp-log* buffer` and as much as possible details for a post-mortem analysis should go to \*lsp-log\*. Some errors occur rarely and reproducing them is difficult. And, for example, you could print the whole bug report message there, so a user could copy-paste it in this form. Make it easy for users to report bugs. 🙂

**Which Language Server did you use**

rust, rust-analyzer (0d03fe6)

**OS**

Linux, Debian/unstable

**Comment by jo-so** [comment]
Sorry, forgot to mention the bad response (I found after some digging through the code):

```
[Trace - 09:45:27 ] Sending request 'textDocument/codeAction - (12)'.
Params: {
  "textDocument": {
    "uri": "file:///home/joerg/Projekte/firma/riafulon/src/user.rs"
  },
  "range": {
    "start": {
      "line": 22,
      "character": 1
    },
    "end": {
      "line": 22,
      "character": 1
    }
  },
  "context": {
    "diagnostics": []
  }
}


[Trace - 09:45:27 ] Sending notification '$/cancelRequest'.
Params: {
  "id": 12
}


[Trace - 09:45:31 ] Received response 'nil - (12)' in 0ms.
Result: null

**Comment by yyoncho** [comment]
This is caused by lsp-mode calling rust-analyzer before it has been initialized(AFAIK there is no way to know that). I think that the same will happen with other clients(vscode, etc) but the difference is that lsp-mode makes the error evident. We get similar errors with ccls which may load its index for a lot of time. I am not sure we could do anything here except allow configuring the way default error handler reports the errors.

**Comment by jo-so** [comment]
> I am not sure we could do anything here except allow configuring the way
> default error handler reports the errors.

That was what I had suspected. That's why I request a better error message.
I don't want you change the behaviour, but improve the message. The current
one doesn't tell the user anything and to report an issue it doesn't provide
much information.

Maybe it's possible to downgrade the message “Rust Analyzer is still
loading...” to a warning or an info message. Or is this text used for all
“unknown” errors?

**Comment by yyoncho** [comment]
The error message comes from the server and it is reported by the default error handler which is common for all async requests. Thus, we cannot change it to info/warning.

**Comment by jo-so** [comment]
Would it be possible to have a blacklist of messages they are downgraded? I
know this is hacky, but it gives a better user experience.

**Comment by yyoncho** [comment]
You can customize the output messages that are displayed by putting an advice over `message` or if you are not happy with that solution we may add extension point for `lsp-mode` specifically. We can add message blacklist but it will be overkill and may lead to adding more and more functionality which in general could live in user's config.

**Comment by jo-so** [comment]
It's not a question if the user can fix it with its config. You are in a
better place to know which messages are garbage and which one are important.
When I get such a garbage message, I have to dig through a high bunch of
web search engine results to find out, I can ignore this message. But you
can do lsp-mode users a service and remove these messages.

**Comment by yyoncho** [comment]
> You are in a better place to know which messages are garbage and which one are important.

I don't think we are. The same message can be garbage and can be important depending on the context of invocation. Also, the messages do change, and keeping a registry on our side which message is important and which is not something we wanna do considering that we support 50+ servers and those messages are just strings and can change at any time. What we can do here is to hide the messages that come from idle tasks in lsp-log without showing them to the user. VScode is doing that but I don't feel that this is the best option for us, since it will hide potential issues and both vscode/emacs have different target groups.
